### ISSUE TITLE
Multiprocess mode is slow

### ISSUE THREAD

**Issue by akx** [prometheus/client_python#367]
We're using uwsgi to run in production with `workers > 1`, and as such, we need to use the client's multiprocess mode.

The "this thing doesn't work in multiprocess mode" features aside, it works acceptably in the beginning. However, presenting metrics gets slower and slower as the container we run in lives on, eventually leading to the Prometheus server timing out scraping metrics. As a stopgap measure, I've just increased the timeout on the server's end, but that's not the best of solutions.

Namely, by the time I have ~6800 .db files (1 MiB each, so roughly 7 gigs altogether) in the multiproc dir, generating 341,361 bytes of metrics takes 40.75 seconds on my dev machine (script below). Yikes! Running under a profiler gives roughly the following, sorted by OwnTime:

Name | Call# | Time (ms) | OwnTime (ms)
-- | -- | -- | --
prometheus_client.mmap_dict.MmapedDict#__init__ | 6802 | 25082 | 24570
prometheus_client.multiprocess.MultiProcessCollector#merge | 1 | 44712 | 4107
_io.BufferedReader#close  | 6802 | 3038 | 3038
prometheus_client.mmap_dict.MmapedDict#_read_all_values | 673199 | 3255 | 1824
json.raw_decode | 666397 | 1848 | 1802
json.decode | 666397 | 4145 | 1055
re.Pattern.match  | 1339663 | 978 | 978
json.loads | 666397 | 5298 | 802
sorted  | 669564 | 729 | 658
prometheus_client.metrics_core.Metric#add_sample | 666397 | 1277 | 630
_struct.unpack_from  | 666397 | 504 | 504

```python
import os
import time

os.environ['prometheus_multiproc_dir'] = '/Users/akx/Desktop/roi_prometheus_multiproc'

import prometheus_client
from prometheus_client import multiprocess

t0 = time.time()
registry = prometheus_client.CollectorRegistry()
multiprocess.MultiProcessCollector(registry)
metrics_page = prometheus_client.generate_latest(registry)
print(len(metrics_page))
print(time.time() - t0)
```

Culling some of the files out makes the script linearly faster (4200 files = 25.72 seconds, 2400 files = 12.5 seconds).

I find it hard to believe other users of the library haven't bumped into this somehow, so I'm wondering if anyone has any mitigation ideas? Otherwise I'm tempted to try my hand at writing a different multiprocess backend, such as SQLite, that might be a little slower at writing but likely much faster at collating the data.

**Comment by akx** [comment]
> It is not safe to remove arbitrary files.

Yeah, I'm aware. This was on my local machine, to try things out.

**Comment by fangjianfeng** [comment]
> That's a lot of process churn, you should either try to reduce it or restart the application and wipe the directory regularly. It is not safe to remove arbitrary files.
> 
> > such as SQLite, that might be a little slower at writing but likely much faster at collating the data.
> 
> I would imagine it'd be substantially slower.

wipe the directory regularly will result in loss of metrics。

**Comment by fangjianfeng** [comment]
> We're using uwsgi to run in production with `workers > 1`, and as such, we need to use the client's multiprocess mode.
> 
> The "this thing doesn't work in multiprocess mode" features aside, it works acceptably in the beginning. However, presenting metrics gets slower and slower as the container we run in lives on, eventually leading to the Prometheus server timing out scraping metrics. As a stopgap measure, I've just increased the timeout on the server's end, but that's not the best of solutions.
> 
> Namely, by the time I have ~6800 .db files (1 MiB each, so roughly 7 gigs altogether) in the multiproc dir, generating 341,361 bytes of metrics takes 40.75 seconds on my dev machine (script below). Yikes! Running under a profiler gives roughly the following, sorted by OwnTime:
> 
> Name	Call#	Time (ms)	OwnTime (ms)
> prometheus_client.mmap_dict.MmapedDict#**init**	6802	25082	24570
> prometheus_client.multiprocess.MultiProcessCollector#merge	1	44712	4107
> _io.BufferedReader#close 	6802	3038	3038
> prometheus_client.mmap_dict.MmapedDict#_read_all_values	673199	3255	1824
> json.raw_decode	666397	1848	1802
> json.decode	666397	4145	1055
> re.Pattern.match 	1339663	978	978
> json.loads	666397	5298	802
> sorted 	669564	729	658
> prometheus_client.metrics_core.Metric#add_sample	666397	1277	630
> _struct.unpack_from 	666397	504	504
> ```python
> import os
> import time
> 
> os.environ['prometheus_multiproc_dir'] = '/Users/akx/Desktop/roi_prometheus_multiproc'
> 
> import prometheus_client
> from prometheus_client import multiprocess
> 
> t0 = time.time()
> registry = prometheus_client.CollectorRegistry()
> multiprocess.MultiProcessCollector(registry)
> metrics_page = prometheus_client.generate_latest(registry)
> print(len(metrics_page))
> print(time.time() - t0)
> ```
> 
> Culling some of the files out makes the script linearly faster (4200 files = 25.72 seconds, 2400 files = 12.5 seconds).
> 
> I find it hard to believe other users of the library haven't bumped into this somehow, so I'm wondering if anyone has any mitigation ideas? Otherwise I'm tempted to try my hand at writing a different multiprocess backend, such as SQLite, that might be a little slower at writing but likely much faster at collating the data.

same promblem。i mitigate this problem:
1.periodically restart uwsgi (when requests >=10000);
2.use uwsgi worker_id instead of pid；

**Comment by fangjianfeng** [comment]
same problem，now i scrape it 6s

![image](https://user-images.githubusercontent.com/19743705/59689229-dd839c00-9211-11e9-8444-30fd28817bfc.png)

**Comment by akx** [comment]
@fangjianfeng Is this on 0.7.0 or earlier? My PR #421 (released in 0.7.0) provides a significant (6x) speed boost.

**Comment by fangjianfeng** [comment]
> @fangjianfeng Is this on 0.7.0 or earlier? My PR #421 (released in 0.7.0) provides a significant (6x) speed boost.

I made this change under version 0.6.0 and now I have upgraded to 0.7.0. This modification still applies.

**Comment by brian-brazil** [comment]
#443 is a better place to discuss this.
